async_fuse/src/memfs/dir.rs

```diff
@@ -1,46 +1,97 @@
 
-#[cfg(target_os = "linux")]
-use libc::{dirent64 as dirent, readdir64_r as readdir_r};
+/// Directory meta-data
+pub struct Dir(NonNull<libc::DIR>);
 
-#[cfg(target_os = "macos")]
-use libc::{dirent, readdir_r};
+unsafe impl Send for Dir {}
+unsafe impl Sync for Dir {}
 
-/// Directory meta-data
-#[derive(Debug)]
-pub struct Dir(ptr::NonNull<libc::DIR>);
+impl Drop for Dir {
+    fn drop(&mut self) {
+        let dirp = self.0.as_ptr();
+        unsafe {
+            let _ = libc::closedir(dirp);
+        }
+    }
+}
 
 impl Dir {
-    /// Converts from a file descriptor, closing it on success or failure.
-    pub fn from_fd(fd: RawFd) -> nix::Result<Self> {
-        let d = unsafe { libc::fdopendir(fd) };
-        ptr::NonNull::new(d).map_or_else(
-            || {
-                let e = nix::Error::last();
-                unsafe { libc::close(fd) };
-                Err(e)
-            },
-            |non_null_ptr| Ok(Self(non_null_ptr)),
-        )
+    /// Calls opendir(3)
+    /// # Errors
+    /// This method will return `io::Error` if the underlying syscalls fails.
+    #[allow(dead_code)]
+    pub fn opendir<P: AsRef<Path>>(path: P) -> io::Result<Self> {
+        let bytes = path.as_ref().as_os_str().as_bytes();
+        with_c_str(bytes, |p| unsafe {
+            let dirname = p.as_ptr();
+            let dirp = libc::opendir(dirname);
+            if dirp.is_null() {
+                return Err(io::Error::last_os_error());
+            }
+            Ok(Self(NonNull::new_unchecked(dirp)))
+        })
+    }
+
+    /// Calls fdopendir(3)
+    /// # Errors
+    /// This method will return `io::Error` if the underlying syscalls fails.
+    /// When an error occurs, closes the `fd` and returns the previous error.
+    /// # Safety
+    /// This function **consumes ownership** of the specified file descriptor.
+    /// The returned object will take responsibility for closing it when the object goes out of scope.
+    pub unsafe fn fdopendir(fd: RawFd) -> io::Result<Self> {
+        let dirp = libc::fdopendir(fd);
+        if dirp.is_null() {
+            let err = io::Error::last_os_error();
+            let _ = libc::close(fd);
+            return Err(err);
+        }
+        Ok(Self(NonNull::new_unchecked(dirp)))
+    }
+
+    /// See [`Dir::fdopendir`]
+    #[allow(dead_code)]
+    pub unsafe fn try_from_raw_fd(fd: RawFd) -> io::Result<Self> {
+        Self::fdopendir(fd)
     }
 }
 
-// `Dir` is safe to pass from one thread to another, as it's not reference-counted.
-unsafe impl Send for Dir {}
+impl IntoIterator for Dir {
+    type Item = io::Result<DirEntry>;
+
+    type IntoIter = ReadDir;
+
+    /// Returns an iterator over the entries within a directory.
+    fn into_iter(self) -> Self::IntoIter {
+        ReadDir {
+            dir: self,
+            end_of_stream: false,
+        }
+    }
+}
 
 /// Directory entry
 #[derive(Debug)]
 pub struct DirEntry {
     /// The i-number of the entry
-    ino: INum,
+    ino: ino_t,
```

Comment:
This is also a raw data, does it affect the safety of the code?

Reply:
`RawFd` is a resource handle type while `ino_t` (`INum`/`ino_t`/`u64`) is an index type.
Handles should be managed carefully. But indices have nothing to do with any resource. They are just numbers.